File: solutions/misc/frequently-asked-questions/detect-cycle-directed-graph.py
"""
Detect Cycle in a Directed Graph
Pattern: DFS with 3-color marking OR Kahn's BFS (topological sort)

Time:  O(V + E)
Space: O(V)
"""

from collections import defaultdict, deque
from typing import List


def has_cycle_dfs(n: int, edges: List[List[int]]) -> bool:
    """DFS 3-color: WHITE=unvisited, GRAY=in-stack, BLACK=done."""
    # Build adjacency list for the directed graph
    graph = defaultdict(list)
    for u, v in edges:
        graph[u].append(v)

    WHITE, GRAY, BLACK = 0, 1, 2
    color = [WHITE] * n  # Track visitation state of each node

    def dfs(node: int) -> bool:
        """Returns True if a cycle is found in the DFS subtree rooted at node."""
        color[node] = GRAY  # Mark node as being in current recursion stack
        for neighbor in graph[node]:
            if color[neighbor] == GRAY:
                # Back edge found: neighbor is already in current path → cycle
                return True
            if color[neighbor] == WHITE and dfs(neighbor):
                # If neighbor's DFS finds a cycle, propagate it up
                return True
        color[node] = BLACK  # Mark node as fully processed (no cycle in its subtree)
        return False

    # Run DFS from each unvisited node (handles disconnected graphs)
    return any(color[i] == WHITE and dfs(i) for i in range(n))


def has_cycle_bfs(n: int, edges: List[List[int]]) -> bool:
    """Kahn's BFS: cycle ↔ not all nodes can be topologically ordered."""
    # Build adjacency list and compute in-degrees
    graph = defaultdict(list)
    in_degree = [0] * n

    for u, v in edges:
        graph[u].append(v)
        in_degree[v] += 1

    # Initialize queue with all nodes having in-degree 0 (no dependencies)
    queue = deque(i for i in range(n) if in_degree[i] == 0)
    processed = 0

    # Process nodes in topological order
    while queue:
        node = queue.popleft()
        processed += 1
        # Remove outgoing edges from node, decrement neighbors' in-degrees
        for neighbor in graph[node]:
            in_degree[neighbor] -= 1
            # If neighbor's in-degree becomes 0, all its dependencies are processed
            if in_degree[neighbor] == 0:
                queue.append(neighbor)

    # If not all nodes processed, there's a cycle (nodes with remaining in-degrees)
    return processed != n


# ── Tests ─────────────────────────────────────────────────────────────────────
if __name__ == "__main__":
    print("DFS cycle:",    has_cycle_dfs(4, [[0,1],[1,2],[2,3],[3,1]]))  # True
    print("DFS no cycle:", has_cycle_dfs(4, [[0,1],[1,2],[2,3]]))        # False
    print("BFS cycle:",    has_cycle_bfs(4, [[0,1],[1,2],[2,3],[3,1]]))  # True
    print("BFS no cycle:", has_cycle_bfs(4, [[0,1],[1,2],[2,3]]))        # False

File: solutions/misc/frequently-asked-questions/detect-cycle-directed-graph.js
/**
 * Detect Cycle in a Directed Graph
 * Pattern: DFS with 3-color marking (WHITE=0, GRAY=1, BLACK=2)
 *
 * Time:  O(V + E)
 * Space: O(V) for color array + recursion stack
 *
 * 60-Second ID:
 *   Core DS: directed graph → adjacency list
 *   Core Op: detect cycle (a back-edge = cycle)
 *   Constraint: directed (can't use simple visited set — needs in-stack tracking)
 *   → DFS with GRAY (in-stack) coloring; GRAY→GRAY edge = cycle
 *
 * Also: Topological sort via Kahn's BFS — if nodes with 0 in-degree
 * don't cover all nodes → cycle exists (alternate approach).
 */

/**
 * DFS 3-color approach
 * @param {number} n - number of nodes (0..n-1)
 * @param {number[][]} edges - [from, to] pairs
 * @returns {boolean} true if cycle exists
 */
function hasCycleDFS(n, edges) {
  const graph = Array.from({ length: n }, () => []);
  for (const [u, v] of edges) graph[u].push(v);

  const WHITE = 0, GRAY = 1, BLACK = 2;
  const color = new Array(n).fill(WHITE);

  function dfs(node) {
    color[node] = GRAY; // mark as in-stack
    for (const neighbor of graph[node]) {
      if (color[neighbor] === GRAY) return true;  // back edge = cycle
      if (color[neighbor] === WHITE && dfs(neighbor)) return true;
    }
    color[node] = BLACK; // fully processed
    return false;
  }

  for (let i = 0; i < n; i++) {
    if (color[i] === WHITE && dfs(i)) return true;
  }
  return false;
}

/**
 * Kahn's BFS topological sort approach (cycle = not all nodes processed)
 * @param {number} n
 * @param {number[][]} edges
 * @returns {boolean}
 */
function hasCycleBFS(n, edges) {
  const graph = Array.from({ length: n }, () => []);
  const inDegree = new Array(n).fill(0);

  for (const [u, v] of edges) {
    graph[u].push(v);
    inDegree[v]++;
  }

  const queue = [];
  for (let i = 0; i < n; i++) {
    if (inDegree[i] === 0) queue.push(i);
  }

  let processed = 0;
  while (queue.length > 0) {
    const node = queue.shift();
    processed++;
    for (const neighbor of graph[node]) {
      inDegree[neighbor]--;
      if (inDegree[neighbor] === 0) queue.push(neighbor);
    }
  }

  return processed !== n; // if not all processed → cycle
}

// ─── Tests ───────────────────────────────────────────────────────────────────
console.log("DFS cycle:", hasCycleDFS(4, [[0,1],[1,2],[2,3],[3,1]])); // true  (3→1 is back edge)
console.log("DFS no cycle:", hasCycleDFS(4, [[0,1],[1,2],[2,3]]));   // false
console.log("BFS cycle:", hasCycleBFS(4, [[0,1],[1,2],[2,3],[3,1]])); // true
console.log("BFS no cycle:", hasCycleBFS(4, [[0,1],[1,2],[2,3]]));   // false

module.exports = { hasCycleDFS, hasCycleBFS };
